ddtrace/appsec/sqreen.py

```diff
@@ -0,0 +1,33 @@
+from collections import Counter
+import logging
+from typing import Any
+from typing import Mapping
+
+from sq_native import waf  # type: ignore
+
+from ddtrace.appsec.protections import BaseProtection
+
+
+log = logging.getLogger(__name__)
+
+
+class SqreenLibrary(BaseProtection):
+    """
+    Process application data with the Sqreen library.
+    """
+
+    budget = 5 * 1000  # max 5ms budget
+
+    def __init__(self, rules):
+        self._instance = waf.WAFEngine(rules)
+        self.stats = Counter({"total": 0, "reported": 0})
+
+    def process(self, context_id, data):
+        # type: (int, Mapping[str, Any]) -> None
+        log.debug("Create a new Sqreen context for %r", context_id)
+        context = self._instance.create_context()
+        ret = context.run(data, self.budget)
```

Comment:
Does `context.run` acquire/require the GIL? If not, perhaps this could run in its own thread

Reply:
Thanks for the review!

`context.run` releases the GIL for the most part of its execution so it makes sense to run it on its own thread but there a few caveats:
* it's hard to scale the number of worker threads to match the load of the HTTP workers. For example, let's say an app uses gevent and can handle 100 concurrent requests with a single process, how many threads should we use?
* there is a small latency penalty induced by synchronisation (queuing the data & waiting for the result). We tried to use `concurrent.futures` but it was really slow on Python 2. Maybe there is another implementation than https://github.com/agronholm/pythonfutures but I didn't move forward on the experimentation.

Reply:
Can't the work be queued up and handled by a single thread at the consuming end of the queue? Of course I'm assuming that what `context.run` does is relatively expensive based on the allowed `budget`, but if this is rather quick on average it's fine to have it block the thread it runs in. Other solution might inevitably introduce more overheads, like a multiprocess approach.

Reply:
Running the expensive processing (e.g. `context.run`) asynchronously is indeed a promising approach and other APMs are also considering it (cc @Anilm3) but let's do the processing synchronously until we figure out a good enough solution. I added a configurable budget (with the `DD_APPSEC_SQREEN_BUDGET_MS` env. variable) and some metrics to get a better idea of the real processing time.